File: app/src/main/java/com/vteam/testdemo/profile/CreateProfileActivity.java
package com.vteam.testdemo.profile;

import android.app.ProgressDialog;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import com.bumptech.glide.Glide;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;
import com.theartofdev.edmodo.cropper.CropImage;
import com.theartofdev.edmodo.cropper.CropImageView;
import com.vteam.testdemo.R;
import com.vteam.testdemo.common.Constants;
import com.vteam.testdemo.landing.LandingActivity;
import com.vteam.testdemo.landing.model.UserStatus;
import com.vteam.testdemo.landing.model.Users;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.HashMap;
import java.util.Map;

import de.hdodenhof.circleimageview.CircleImageView;

public class CreateProfileActivity extends AppCompatActivity {


    private static final int GALLERY_PICK = 1;
    private DatabaseReference RootRef;
    private FirebaseAuth mAuth;
    private String currentUserID;
    private Button UpdateAccountSettings;
    private EditText userName, userStatus;
    private CircleImageView userProfileImage;
    private StorageReference UserProfileImagesRef;
    private ProgressDialog loadingBar;
    private Toolbar SettingsToolBar;
    private CircleImageView userProfileImageEdit;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_settings);
        setActionBar();
        InitializeFields();
        mAuth = FirebaseAuth.getInstance();
        currentUserID = mAuth.getCurrentUser().getUid();
        RootRef = FirebaseDatabase.getInstance().getReference();
        UserProfileImagesRef = FirebaseStorage.getInstance().getReference().child("Profile Images");

//        userName.setEnabled(false);

        UpdateAccountSettings.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                UpdateSettings();
            }
        });

        RetrieveUserInfo();

        userProfileImageEdit.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
//                Intent galleryIntent = new Intent();
//                galleryIntent.setAction(Intent.ACTION_GET_CONTENT);
//                galleryIntent.setType("image/*");
//                startActivityForResult(galleryIntent, GALLERY_PICK);
                CropImage.activity()
                        .setGuidelines(CropImageView.Guidelines.ON)
                        .setAspectRatio(1, 1)
                        .start(CreateProfileActivity.this);
            }
        });
    }

    private void updateUserStatus(String state) {
        String saveCurrentTime, saveCurrentDate;

        Calendar calendar = Calendar.getInstance();


        SimpleDateFormat currentDate = new SimpleDateFormat(Constants.PATTERN.PATTERN_MMM_DD_yyyy);
        saveCurrentDate = currentDate.format(calendar.getTime());


        SimpleDateFormat currentTime = new SimpleDateFormat(Constants.PATTERN.PATTERN_hh_mm_a);
        saveCurrentTime = currentTime.format(calendar.getTime());

        final HashMap<String, Object> onlineStateMap = new HashMap<>();
        onlineStateMap.put("time", saveCurrentTime);
        onlineStateMap.put("date", saveCurrentDate);
        onlineStateMap.put("state", state);

//
//        RootRef.child("Users").child(currentUserID).child("userState")
//                            .updateChildren(onlineStateMap);


    }


    private void UpdateSettings() {
        String userName = this.userName.getText().toString();
        String status = userStatus.getText().toString();
        if (TextUtils.isEmpty(userName)) {
            Toast.makeText(this, "Please write your user name first....", Toast.LENGTH_SHORT).show();
        }
        if (TextUtils.isEmpty(status)) {
            Toast.makeText(this, "Please write your status....", Toast.LENGTH_SHORT).show();
        } else {

            Users users = new Users(currentUserID);
            users.setName(userName);
            users.setStatus(status);
//            HashMap<String, Object> profileMap = new HashMap<>();
//            profileMap.put("uid", currentUserID);
//            profileMap.put("name", userName);
//            profileMap.put("status", status);



            String saveCurrentTime, saveCurrentDate;

            Calendar calendar = Calendar.getInstance();

            SimpleDateFormat currentDate = new SimpleDateFormat(Constants.PATTERN.PATTERN_MMM_DD_yyyy);
            saveCurrentDate = currentDate.format(calendar.getTime());

            SimpleDateFormat currentTime = new SimpleDateFormat(Constants.PATTERN.PATTERN_hh_mm_a);
            saveCurrentTime = currentTime.format(calendar.getTime());

            UserStatus userStatus = new UserStatus();
            userStatus.setDate(saveCurrentDate);
            userStatus.setTime(saveCurrentTime);
            userStatus.setStatus(Constants.USER_STATE.ONLINE);

            users.setUserStatus(userStatus);
            Map<String, Object> userMap = new HashMap<>();
            userMap.put(currentUserID,users);


            RootRef.child("Users").updateChildren(userMap).addOnCompleteListener(new OnCompleteListener<Void>() {
                @Override
                public void onComplete(@NonNull Task<Void> task) {
                    if (task.isSuccessful()) {
                        SendUserToMainActivity();
                        Toast.makeText(CreateProfileActivity.this, "Profile Updated Successfully...", Toast.LENGTH_SHORT).show();
                    } else {
                        String message = task.getException().toString();
                        Toast.makeText(CreateProfileActivity.this, "Error: " + message, Toast.LENGTH_SHORT).show();
                    }
                }
            });
        }
    }

    private void RetrieveUserInfo() {
        RootRef.child("Users").child(currentUserID).addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                if ((dataSnapshot.exists()) && (dataSnapshot.hasChild("name") && (dataSnapshot.hasChild("image")))) {
                    String retrieveUserName = dataSnapshot.child("name").getValue().toString();
                    String retrieveStatus = dataSnapshot.child("status").getValue().toString();
                    String retrieveProfilePhoto = dataSnapshot.child("image").getValue().toString();

                    userName.setText(retrieveUserName);
                    userStatus.setText(retrieveStatus);
                    StorageReference reference = FirebaseStorage.getInstance().getReference();

                    reference.child(retrieveProfilePhoto).getDownloadUrl().addOnSuccessListener(new OnSuccessListener<Uri>() {
                        @Override
                        public void onSuccess(Uri uri) {
                            // Got the download URL for 'users/me/profile.png'
                            Log.d("URL", "" + uri);
                            Glide.with(getApplicationContext())
                                    .load(uri)
                                    .into(userProfileImage);
                        }
                    }).addOnFailureListener(new OnFailureListener() {
                        @Override
                        public void onFailure(@NonNull Exception exception) {
                            // Handle any errors
                        }
                    });
                } else if ((dataSnapshot.exists()) && (dataSnapshot.hasChild("name"))) {
                    String retrieveUserName = dataSnapshot.child("name").getValue().toString();
                    String retrievesStatus = dataSnapshot.child("status").getValue().toString();

                    userName.setText(retrieveUserName);
                    userStatus.setText(retrievesStatus);
                } else {
                    userName.setVisibility(View.VISIBLE);
                    Toast.makeText(CreateProfileActivity.this, "Please set & update your profile information...", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }

    private void SendUserToMainActivity() {
        Intent mainIntent = new Intent(CreateProfileActivity.this, LandingActivity.class);
        mainIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        startActivity(mainIntent);
        finish();
    }

    private void InitializeFields() {
        UpdateAccountSettings = (Button) findViewById(R.id.update_settings_button);
        userName = (EditText) findViewById(R.id.set_user_name);
        userStatus = (EditText) findViewById(R.id.set_profile_status);
        userProfileImage = (CircleImageView) findViewById(R.id.set_profile_image);
        userProfileImageEdit = findViewById(R.id.profile_image_edit);
        loadingBar = new ProgressDialog(this);


    }

    private void setActionBar() {
        SettingsToolBar = (Toolbar) findViewById(R.id.settings_toolbar);
        setSupportActionBar(SettingsToolBar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowCustomEnabled(true);
        getSupportActionBar().setTitle("Profile Update");
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == GALLERY_PICK) {
            if (resultCode == RESULT_OK && data != null) {
                Uri ImageUri = data.getData();

                CropImage.activity()
                        .setGuidelines(CropImageView.Guidelines.ON)
                        .setAspectRatio(1, 1)
                        .start(this);
            }
        } else if (requestCode == CropImage.CROP_IMAGE_ACTIVITY_REQUEST_CODE) {
            CropImage.ActivityResult result = CropImage.getActivityResult(data);

            if (resultCode == RESULT_OK) {
                loadingBar.setTitle("Set Profile Image");
                loadingBar.setMessage("Please wait, your profile image is updating...");
                loadingBar.setCanceledOnTouchOutside(false);
                loadingBar.show();

                Uri resultUri = result.getUri();


                StorageReference filePath = UserProfileImagesRef.child(currentUserID + ".jpg");

                filePath.putFile(resultUri).addOnCompleteListener(new OnCompleteListener<UploadTask.TaskSnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<UploadTask.TaskSnapshot> task) {
                        if (task.isSuccessful()) {
                            Toast.makeText(CreateProfileActivity.this, "Profile Image uploaded Successfully...", Toast.LENGTH_SHORT).show();

                            final String downloaedUrl = task.getResult().getMetadata().getPath();

                            RootRef.child("Users").child(currentUserID).child("image")
                                    .setValue(downloaedUrl)
                                    .addOnCompleteListener(new OnCompleteListener<Void>() {
                                        @Override
                                        public void onComplete(@NonNull Task<Void> task) {
                                            if (task.isSuccessful()) {
                                                Toast.makeText(CreateProfileActivity.this, "Image save in Database, Successfully...", Toast.LENGTH_SHORT).show();
                                                loadingBar.dismiss();
                                            } else {
                                                String message = task.getException().toString();
                                                Toast.makeText(CreateProfileActivity.this, "Error: " + message, Toast.LENGTH_SHORT).show();
                                                loadingBar.dismiss();
                                            }
                                        }
                                    });
                        } else {
                            String message = task.getException().toString();
                            Toast.makeText(CreateProfileActivity.this, "Error: " + message, Toast.LENGTH_SHORT).show();
                            loadingBar.dismiss();
                        }
                    }
                });
            }
        }
    }


}

File: app/src/main/java/com/vteam/testdemo/profile/CreateProfileActivity.kt
package com.vteam.testdemo.profile

import android.app.Activity
import android.app.ProgressDialog
import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import com.bumptech.glide.Glide
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageReference
import com.theartofdev.edmodo.cropper.CropImage
import com.theartofdev.edmodo.cropper.CropImageView
import com.vteam.testdemo.R
import com.vteam.testdemo.common.Constants
import com.vteam.testdemo.landing.LandingActivity
import com.vteam.testdemo.landing.model.UserStatus
import com.vteam.testdemo.landing.model.Users
import com.vteam.testdemo.profile.CreateProfileActivity
import de.hdodenhof.circleimageview.CircleImageView
import java.text.SimpleDateFormat
import java.util.*

class CreateProfileActivity : AppCompatActivity() {
    private var RootRef: DatabaseReference? = null
    private var mAuth: FirebaseAuth? = null
    private var currentUserID: String? = null
    private var UpdateAccountSettings: Button? = null
    private var userName: EditText? = null
    private var userStatus: EditText? = null
    private var userProfileImage: CircleImageView? = null
    private var UserProfileImagesRef: StorageReference? = null
    private var loadingBar: ProgressDialog? = null
    private var SettingsToolBar: Toolbar? = null
    private var userProfileImageEdit: CircleImageView? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_settings)
        setActionBar()
        InitializeFields()
        mAuth = FirebaseAuth.getInstance()
        currentUserID = mAuth!!.currentUser!!.uid
        RootRef = FirebaseDatabase.getInstance().reference
        UserProfileImagesRef = FirebaseStorage.getInstance().reference.child("Profile Images")

//        userName.setEnabled(false);
        UpdateAccountSettings!!.setOnClickListener { UpdateSettings() }
        RetrieveUserInfo()
        userProfileImageEdit!!.setOnClickListener { //                Intent galleryIntent = new Intent();
//                galleryIntent.setAction(Intent.ACTION_GET_CONTENT);
//                galleryIntent.setType("image/*");
//                startActivityForResult(galleryIntent, GALLERY_PICK);
            CropImage.activity()
                .setGuidelines(CropImageView.Guidelines.ON)
                .setAspectRatio(1, 1)
                .start(this@CreateProfileActivity)
        }
    }

    private fun updateUserStatus(state: String) {
        val saveCurrentTime: String
        val saveCurrentDate: String
        val calendar = Calendar.getInstance()
        val currentDate =
            SimpleDateFormat(Constants.PATTERN.PATTERN_MMM_DD_yyyy)
        saveCurrentDate = currentDate.format(calendar.time)
        val currentTime =
            SimpleDateFormat(Constants.PATTERN.PATTERN_hh_mm_a)
        saveCurrentTime = currentTime.format(calendar.time)
        val onlineStateMap =
            HashMap<String, Any>()
        onlineStateMap["time"] = saveCurrentTime
        onlineStateMap["date"] = saveCurrentDate
        onlineStateMap["state"] = state

//
//        RootRef.child("Users").child(currentUserID).child("userState")
//                            .updateChildren(onlineStateMap);
    }

    private fun UpdateSettings() {
        val userName = userName!!.text.toString()
        val status = userStatus!!.text.toString()
        if (TextUtils.isEmpty(userName)) {
            Toast.makeText(this, "Please write your user name first....", Toast.LENGTH_SHORT).show()
        }
        if (TextUtils.isEmpty(status)) {
            Toast.makeText(this, "Please write your status....", Toast.LENGTH_SHORT).show()
        } else {
            val users = Users(currentUserID)
            users.name = userName
            users.status = status
            //            HashMap<String, Object> profileMap = new HashMap<>();
//            profileMap.put("uid", currentUserID);
//            profileMap.put("name", userName);
//            profileMap.put("status", status);
            val saveCurrentTime: String
            val saveCurrentDate: String
            val calendar = Calendar.getInstance()
            val currentDate =
                SimpleDateFormat(Constants.PATTERN.PATTERN_MMM_DD_yyyy)
            saveCurrentDate = currentDate.format(calendar.time)
            val currentTime =
                SimpleDateFormat(Constants.PATTERN.PATTERN_hh_mm_a)
            saveCurrentTime = currentTime.format(calendar.time)
            val userStatus = UserStatus()
            userStatus.date = saveCurrentDate
            userStatus.time = saveCurrentTime
            userStatus.status = Constants.USER_STATE.ONLINE
            users.userStatus = userStatus
            val userMap: MutableMap<String?, Any> =
                HashMap()
            userMap[currentUserID] = users
            RootRef!!.child(Constants.NODES.USER_NODE)
                .updateChildren(userMap)
                .addOnCompleteListener { task ->
                    if (task.isSuccessful) {
                        SendUserToMainActivity()
                        Toast.makeText(
                            this@CreateProfileActivity,
                            "Profile Updated Successfully...",
                            Toast.LENGTH_SHORT
                        ).show()
                    } else {
                        val message = task.exception.toString()
                        Toast.makeText(
                            this@CreateProfileActivity,
                            "Error: $message",
                            Toast.LENGTH_SHORT
                        ).show()
                    }
                }
        }
    }

    private fun RetrieveUserInfo() {
        RootRef!!.child(Constants.NODES.USER_NODE).child(currentUserID!!)
            .addValueEventListener(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    if (dataSnapshot.exists() && dataSnapshot.hasChild("name") && dataSnapshot.hasChild(
                            "image"
                        )
                    ) {
                        val retrieveUserName =
                            dataSnapshot.child("name").value.toString()
                        val retrieveStatus =
                            dataSnapshot.child("status").value.toString()
                        val retrieveProfilePhoto =
                            dataSnapshot.child("image").value.toString()
                        userName!!.setText(retrieveUserName)
                        userStatus!!.setText(retrieveStatus)
                        val reference =
                            FirebaseStorage.getInstance().reference
                        reference.child(retrieveProfilePhoto).downloadUrl
                            .addOnSuccessListener { uri -> // Got the download URL for 'users/me/profile.png'
                                Log.d("URL", "" + uri)
                                Glide.with(applicationContext)
                                    .load(uri)
                                    .into(userProfileImage!!)
                            }.addOnFailureListener {
                                // Handle any errors
                            }
                    } else if (dataSnapshot.exists() && dataSnapshot.hasChild("name")) {
                        val retrieveUserName =
                            dataSnapshot.child("name").value.toString()
                        val retrievesStatus =
                            dataSnapshot.child("status").value.toString()
                        userName!!.setText(retrieveUserName)
                        userStatus!!.setText(retrievesStatus)
                    } else {
                        userName!!.visibility = View.VISIBLE
                        Toast.makeText(
                            this@CreateProfileActivity,
                            "Please set & update your profile information...",
                            Toast.LENGTH_SHORT
                        ).show()
                    }
                }

                override fun onCancelled(databaseError: DatabaseError) {}
            })
    }

    private fun SendUserToMainActivity() {
        val mainIntent = Intent(this@CreateProfileActivity, LandingActivity::class.java)
        mainIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
        startActivity(mainIntent)
        finish()
    }

    private fun InitializeFields() {
        UpdateAccountSettings =
            findViewById<View>(R.id.update_settings_button) as Button
        userName = findViewById<View>(R.id.set_user_name) as EditText
        userStatus = findViewById<View>(R.id.set_profile_status) as EditText
        userProfileImage =
            findViewById<View>(R.id.set_profile_image) as CircleImageView
        userProfileImageEdit =
            findViewById(R.id.profile_image_edit)
        loadingBar = ProgressDialog(this)
    }

    private fun setActionBar() {
        SettingsToolBar =
            findViewById<View>(R.id.settings_toolbar) as Toolbar
        setSupportActionBar(SettingsToolBar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.setDisplayShowCustomEnabled(true)
        supportActionBar!!.title = "Profile Update"
    }

    override fun onActivityResult(
        requestCode: Int,
        resultCode: Int,
        data: Intent?
    ) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == GALLERY_PICK) {
            if (resultCode == Activity.RESULT_OK && data != null) {
                val ImageUri = data.data
                CropImage.activity()
                    .setGuidelines(CropImageView.Guidelines.ON)
                    .setAspectRatio(1, 1)
                    .start(this)
            }
        } else if (requestCode == CropImage.CROP_IMAGE_ACTIVITY_REQUEST_CODE) {
            val result = CropImage.getActivityResult(data)
            if (resultCode == Activity.RESULT_OK) {
                loadingBar!!.setTitle("Set Profile Image")
                loadingBar!!.setMessage("Please wait, your profile image is updating...")
                loadingBar!!.setCanceledOnTouchOutside(false)
                loadingBar!!.show()
                val resultUri = result.uri
                val filePath = UserProfileImagesRef!!.child("$currentUserID.jpg")
                filePath.putFile(resultUri).addOnCompleteListener { task ->
                    if (task.isSuccessful) {
                        Toast.makeText(
                            this@CreateProfileActivity,
                            "Profile Image uploaded Successfully...",
                            Toast.LENGTH_SHORT
                        ).show()
                        val downloaedUrl =
                            task.result.metadata!!.path
                        RootRef!!.child(Constants.NODES.USER_NODE)
                            .child(currentUserID!!).child("image")
                            .setValue(downloaedUrl)
                            .addOnCompleteListener { task ->
                                if (task.isSuccessful) {
                                    Toast.makeText(
                                        this@CreateProfileActivity,
                                        "Image save in Database, Successfully...",
                                        Toast.LENGTH_SHORT
                                    ).show()
                                    loadingBar!!.dismiss()
                                } else {
                                    val message =
                                        task.exception.toString()
                                    Toast.makeText(
                                        this@CreateProfileActivity,
                                        "Error: $message",
                                        Toast.LENGTH_SHORT
                                    ).show()
                                    loadingBar!!.dismiss()
                                }
                            }
                    } else {
                        val message = task.exception.toString()
                        Toast.makeText(
                            this@CreateProfileActivity,
                            "Error: $message",
                            Toast.LENGTH_SHORT
                        ).show()
                        loadingBar!!.dismiss()
                    }
                }
            }
        }
    }

    companion object {
        const val GALLERY_PICK = 1
    }
}